enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,458 @@
+
+
+## Design Details for `Auto discovery of local devices`
+
+API scheme for `LocalVolumeDiscovery` and `LocalVolumeNodeDiscovery` CR:
+
+```go
+type LocalVolumeDiscoveryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeDiscovery `json:"items"`
+}
+
+type LocalVolumeDiscovery struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeDiscoverySpec   `json:"spec"`
+	Status            LocalVolumeDiscoveryStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeDiscoverySpec struct {
+	// Nodes on which the automatic detection policies must run.
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscovery Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+type LocalVolumeDiscoveryStatus struct {
+	Phase string `json:"phase,omitempty"`
+	// RelatedObjects is a list of objects created and maintained by this
+	// operator. Object references will be added to this list after they have
+	// been created AND found in the cluster.
+	// +optional
+	RelatedObjects []corev1.ObjectReference `json:"relatedObjects,omitempty"`
+}
+
+type LocalVolumeNodeDiscoveryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeNodeDiscovery `json:"items"`
+}
+
+type LocalVolumeNodeDiscovery struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              NodeDeviceMapSpec   `json:"spec"`
+	Status            NodeDeviceMapStatus `json:"status,omitempty"`
+}
+
+type NodeDeviceMapSpec struct {
+	NodeName string `json:"nodeName"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscovery Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+type NodeDeviceMapStatus struct {
+	Phase      string `json:"phase,omitempty"`
+	TimeStramp string `json:"timestramp,omitempty"`
+	// Blockdevices will be containing the output of following command unmarshalled into the structured
+	// "lsblk -o name,model,vendor,kname,pkname,type,uuid,tran,ro,rm,parttype,serial,size,rota,mountpoint"
+	Blockdevices []Device `json:"blockdevices"`
+}
+
+type Device struct {
+	Name       string   `json:"name,omitempty"`
+	Model      string   `json:"model,omitempty"`
+	Vendor     string   `json:"vendor,omitempty"`
+	Kname      string   `json:"kname,omitempty"`
+	Pkname     string   `json:"pkname,omitempty"`
+	Type       string   `json:"type,omitempty"`
+	UUID       string   `json:"uuid,omitempty"`
+	Tran       string   `json:"tran,omitempty"`
```

Comment:
Can we do a minimum one line documentation of these fields and which is mandatory and which is optional?

Reply:
Are you suggesting to define which properties we require to discover on a device or else consider it a failure? For example, we better discover name, ro, type, and size, (and maybe other fields) or else it's not going to be a useful device.

Reply:
These properties are properties of `lsblk`, how about pointing it lsblk doc ?

Reply:
I'd suggest we decouple the lsblk output from the properties in the CR status. This way we can name the properties to be something more useful, we can use other tools to add other properties (like filesystem), and we can have flexibility to change types (`int`) if needed. This means we just need to define two types: 1) to deserialize the lsblk output and 2) for the CR properties.

Reply:
since the lsblk output is in string we can hash it, and store it in the CR.

Reply:
@travisn I meant this `Device` CR needs bit more elaboration. Currently it looks like every field is optional if so what is the point of creating an empty `Device` object?  

We need to follow k8s API conventions and define at minimum which field must be present in `Device` struct and which ones cant be omitted. As it stands right now - it is very sloppy design.